src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+		    });
+	} else {
+		assert(tsk.get_execution_target() == execution_target::host);
+		assert(chunk.memory_id == host_memory_id);
+		assert(reduction_map.empty());
+		return create<host_task_instruction>(command_batch, tsk.get_launcher<host_task_launcher>(), chunk.execution_range, tsk.get_global_size(),
+		    std::move(allocation_map), tsk.get_collective_group_id() CELERITY_DETAIL_IF_ACCESSOR_BOUNDARY_CHECK(, tsk.get_id(), tsk.get_debug_name()),
+		    [&](const auto& record_debug_info) { record_debug_info(ecmd.get_tid(), ecmd.get_cid(), tsk.get_debug_name(), buffer_memory_access_map); });
+	}
+}
+
+void generator_impl::perform_task_buffer_accesses(
+    const task& tsk, const std::vector<localized_chunk>& concurrent_chunks, const std::vector<instruction*>& command_instructions) //
+{
+	const auto& bam = tsk.get_buffer_access_map();
+	if(bam.get_num_accesses() == 0 && tsk.get_reductions().empty()) return;
+
+	// 1. Collect the read-sets and write-sets of all concurrent chunks on all buffers (TODO this is what buffer_access_map should actually return)
+
+	struct read_write_sets {
+		region<3> reads;
+		region<3> writes;
+	};
+
+	std::vector<std::unordered_map<buffer_id, read_write_sets>> concurrent_read_write_sets(concurrent_chunks.size());
+
+	for(const auto bid : bam.get_accessed_buffers()) {
+		for(size_t i = 0; i < concurrent_chunks.size(); ++i) {
+			read_write_sets rw;
+			for(const auto mode : bam.get_access_modes(bid)) {
+				const auto req =
+				    bam.get_mode_requirements(bid, mode, tsk.get_dimensions(), concurrent_chunks[i].execution_range.get_subrange(), tsk.get_global_size());
+				if(access::mode_traits::is_consumer(mode)) { rw.reads = region_union(rw.reads, req); }
+				if(access::mode_traits::is_producer(mode)) { rw.writes = region_union(rw.writes, req); }
+			}
+			concurrent_read_write_sets[i].emplace(bid, std::move(rw));
+		}
+	}
+
+	for(const auto& rinfo : tsk.get_reductions()) {
+		for(size_t i = 0; i < concurrent_chunks.size(); ++i) {
+			auto& rw_map = concurrent_read_write_sets[i][rinfo.bid]; // allow default-insert on `bid`
+			rw_map.writes = region_union(rw_map.writes, scalar_reduction_box);
+		}
+	}
+
+	// 2. Insert all true-dependencies for reads and anti-dependencies for writes. We do this en-bloc instead of using `perform_concurrent_read_from_allocation`
+	// or `perform_atomic_write_to_allocation` to avoid incorrect dependencies between our concurrent chunks by updating tracking structures too early.
+
+	for(size_t i = 0; i < concurrent_chunks.size(); ++i) {
+		for(const auto& [bid, rw] : concurrent_read_write_sets[i]) {
+			auto& buffer = m_buffers.at(bid);
+			auto& memory = buffer.memories[concurrent_chunks[i].memory_id];
+
+			for(auto& allocation : memory.allocations) {
+				add_dependencies_on_last_writers(
+				    command_instructions[i], allocation, region_intersection(rw.reads, allocation.box), instruction_dependency_origin::read_from_allocation);
+				add_dependencies_on_last_concurrent_accesses(
```

Comment:
If `rw.reads == rw.writes`, then yes. A dependency is also duplicated if it appears between multiple accessed allocations.

This should not affect correctness but it might be worthwhile to discuss the performance impact of requiring the executor to sort through multiple identical dependencies vs. using something like an `unordered_set` of dependencies in `instruction`.